### PR TITLE
DAOS-6604 coverity: defect in fs_copy

### PR DESCRIPTION
**Pull request by daltonbohning**
Move an rc check + free to fix resource leak.

Features: fs_copy,-hw

Signed-off-by: Dalton Bohning <daltonx.bohning@intel.com>

### CODE REVIEW

File: src/utils/daos_hdlr.c [hunk 1]
```diff
@@ -2554,6 +2554,11 @@ fs_copy(struct file_dfs *src_file_dfs,
 	if (dst_file_dfs->obj != NULL || dst_file_dfs->fd != -1)
 		file_close(dst_file_dfs, filename);
 out:
+	if (rc != 0) {
```
Comment (jolivier23) [inline src/utils/daos_hdlr.c:2557]: I don't think this is the right fix.  I think we need to make sure we don't overwrite a failure rc with a success rc on line 2566.  So this gets rid of the memory leak but doesn't actually fix the bug that caused it.

**Review by jolivier23 — CHANGES_REQUESTED** [review]

**Comment by brianjmurrell** [comment]
> Features: fs_copy,-hw

Curious why the `-hw` @daltonbohning.

File: src/utils/daos_hdlr.c (hunk 1 above)
Comment (daltonbohning) [reply src/utils/daos_hdlr.c:2557]: I see. How about this pattern?
```
int close_rc;

close_rc = file_closedir(src_file_dfs, src_dir);
if (close_rc != 0) {
        fprintf(stderr, "Could not close '%s': %d\n",
                   dir_name, close_rc);
        if (rc == 0)
                rc = close_rc;
}
```
FWIW, I have a refactor PR that simplifies this altogether by refactoring individual parts of this function into separate functions.

**Comment by daltonbohning** [comment]
> > Features: fs_copy,-hw
> 
> Curious why the `-hw` @daltonbohning.

Since there are no unit tests for this - only functional - I wanted _**something**_ as a sanity check. Both the vm and hw fs_copy tests will traverse this code (but ideally neither hit the bug), so I ran the cheaper test: vm.

**Comment by brianjmurrell** [comment]
> > > Features: fs_copy,-hw
> > 
> > 
> > Curious why the `-hw` @daltonbohning.
> 
> Since there are no unit tests for this - only functional - I wanted _**something**_ as a sanity check. Both the vm and hw fs_copy tests will traverse this code (but ideally neither hit the bug), so I ran the cheaper test: vm.

Hrm.  I guess I could see that.  But given that you are running the HW stage anyway, I would think the coverage of the test outweighs any (relatively) minor savings in not running that test on HW.  Unless it's a really expensive test on HW.  But even then, I'd suggest that running it is the spirit of what you use `Features:` for anyway: to get the most coverage you can for the thing you are modifying.  The cost of landing a regression is huge.

Anyway, not saying the `-hw` was wrong or anything.  Just offering some points for thought.

**Comment by daltonbohning** [comment]
> Hrm. I guess I could see that. But given that you are running the HW stage anyway, I would think the coverage of the test outweighs any (relatively) minor savings in not running that test on HW. Unless it's a really expensive test on HW. But even then, I'd suggest that running it is the spirit of what you use `Features:` for anyway: to get the most coverage you can for the thing you are modifying. The cost of landing a regression is huge.
> 
> Anyway, not saying the `-hw` was wrong or anything. Just offering some points for thought.

I see. I'll just run `fs_copy` without `-hw` next push, since I agree with this. Although, I can't use `Features` because some of the `hw` tests are in `full_regression`. TBH, I've generally been waiting for `full_regression` to be added to `Features` before using it.

File: src/utils/daos_hdlr.c (hunk 1 above)
Comment (daltonbohning) [reply src/utils/daos_hdlr.c:2557]: I pushed a commit for this so you can see it in code

**Comment by brianjmurrell** [comment]
> I can't use `Features` because some of the `hw` tests are in `full_regression`.

Yeah, I am pretty eager to get that resolved also.  At this point I don't think it's a lot of work.  I just made a comment on the Built and Test channel about it in fact.

**Review by jolivier23 — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]
I actually wonder if we should just log the file_closesdir failure and ignore it.   If we get there, it means the operation to that point has succeeded?

**Comment by daltonbohning** [comment]
> I actually wonder if we should just log the file_closesdir failure and ignore it. If we get there, it means the operation to that point has succeeded?

That particular block of code calls `file_closesdir` on both error conditions and success conditions. So, if we get there and `rc == 0`, then the operation was successful up to that point. Otherwise, if `rc != 0`, then the operations was unsuccessful as some point.

Ideally, this call will probably never fail under normal conditions. We could just log it and ignore, but then I worry that we might ignore something that is important.

**Review by dsikich — APPROVED** [review]

**Pull request merged**